### PR TITLE
TMA-832: support for pluggable component process type

### PR DESCRIPTION
**Pull request by kubamahnert**

### CODE REVIEW

File: lib/gooddata/models/process.rb [hunk 1]
```diff
@@ -90,16 +90,24 @@ def upload_package(path, files_to_exclude, opts = { :client => GoodData.connecti
       # @option options [String] :name Readable name of the process
       # @option options [String] :process_id ID of a process to be redeployed (do not set if you want to create a new process)
       # @option options [Boolean] :verbose (false) Switch on verbose mode for detailed logging
-      def deploy(path, options = { :client => GoodData.client, :project => GoodData.project })
-        return deploy_brick(path, options) if path.to_s.start_with?(APP_STORE_URL)
-
-        return deploy_from_appstore(path.to_s, options) if (path.to_s =~ %r{\${.*}:(.*)\/(.*):\/}) == 0 # rubocop:disable Style/NumericPredicate
+      def deploy(path, options = { client: GoodData.client, project: GoodData.project })
```
Comment (yenkins) [inline lib/gooddata/models/process.rb]: Takes parameters ['options', 'path'] to 5 methods (DataClump)

File: lib/gooddata/models/process.rb [hunk 2]
```diff
@@ -90,16 +90,24 @@ def upload_package(path, files_to_exclude, opts = { :client => GoodData.connecti
...
+        case path
+        when (path.is_a?(Hash) && path[:component])
+          deploy_component path, options
+        when path.to_s.start_with?(APP_STORE_URL)
```
Comment (yenkins) [inline lib/gooddata/models/process.rb]: Calls 'path.to_s' 3 times (DuplicateMethodCall)

File: lib/gooddata/models/process.rb [hunk 3]
```diff
@@ -191,26 +181,43 @@ def deploy_from_appstore(path, options = { :client => GoodData.client, :project
           }
         }
 
+        save(data, options)
+      end
+
+      def deploy_component(data, options = { client: GoodData.client, project: GoodData.project })
+        client, project = GoodData.get_client_and_project(options)
+        data = { process: data } unless data[:process]
```
Comment (yenkins) [inline lib/gooddata/models/process.rb]: Calls 'data[:process]' 4 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/models/process.rb]: Refers to 'data' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/models/process.rb [hunk 4]
```diff
@@ -90,16 +90,24 @@ def upload_package(path, files_to_exclude, opts = { :client => GoodData.connecti
       # @option options [String] :name Readable name of the process
       # @option options [String] :process_id ID of a process to be redeployed (do not set if you want to create a new process)
       # @option options [Boolean] :verbose (false) Switch on verbose mode for detailed logging
-      def deploy(path, options = { :client => GoodData.client, :project => GoodData.project })
-        return deploy_brick(path, options) if path.to_s.start_with?(APP_STORE_URL)
-
-        return deploy_from_appstore(path.to_s, options) if (path.to_s =~ %r{\${.*}:(.*)\/(.*):\/}) == 0 # rubocop:disable Style/NumericPredicate
+      def deploy(path, options = { client: GoodData.client, project: GoodData.project })
+        case path
+        when (path.is_a?(Hash) && path[:component])
```
Comment (yenkins) [inline lib/gooddata/models/process.rb]: Refers to 'path' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/models/process.rb [hunk 5]
```diff
@@ -191,26 +181,43 @@ def deploy_from_appstore(path, options = { :client => GoodData.client, :project
           }
         }
 
+        save(data, options)
+      end
+
+      def deploy_component(data, options = { client: GoodData.client, project: GoodData.project })
```
Comment (yenkins) [inline lib/gooddata/models/process.rb]: Has approx 8 statements (TooManyStatements)

File: lib/gooddata/models/process.rb [hunk 6]
```diff
@@ -90,16 +90,24 @@ def upload_package(path, files_to_exclude, opts = { :client => GoodData.connecti
...
+        when path.to_s.start_with?(APP_STORE_URL)
+          deploy_brick path, options
+        when path.to_s !~ %r{\${.*}:(.*)\/(.*):\/}
+          deploy_from_appstore path.to_s, options
+        else
+          deploy_simple_process path, options
+        end
+      end
 
+      def deploy_simple_process(path, options = { client: GoodData.client, project: GoodData.project })
```
Comment (yenkins) [inline lib/gooddata/models/process.rb]: Has approx 13 statements (TooManyStatements)

File: lib/gooddata/models/process.rb [hunk 7]
```diff
@@ -191,26 +181,43 @@ def deploy_from_appstore(path, options = { :client => GoodData.client, :project
...
+                      key: 'enableEtlComponent',
+                      value: true
+                    }
+
+        save(data, options)
+      end
+
+      private
+
+      def save(data, options = { client: GoodData.client, project: GoodData.project })
```
Comment (yenkins) [inline lib/gooddata/models/process.rb]: Has approx 6 statements (TooManyStatements)

File: lib/gooddata/models/process.rb [hunk 8]
```diff
@@ -191,26 +181,43 @@ def deploy_from_appstore(path, options = { :client => GoodData.client, :project
           }
         }
 
+        save(data, options)
+      end
+
+      def deploy_component(data, options = { client: GoodData.client, project: GoodData.project })
+        client, project = GoodData.get_client_and_project(options)
+        data = { process: data } unless data[:process]
+        data[:process] = GoodData::Helpers.symbolize_keys(data[:process]).select { |k| [:type, :name, :component].include? k }
```
Comment (yenkins) [inline lib/gooddata/models/process.rb]: Has the variable name 'k' (UncommunicativeVariableName)

File: lib/gooddata/models/project.rb [hunk 9]
```diff
@@ -264,6 +264,9 @@ def transfer_processes(from_project, to_project, options = {})
           to_process = if process.path
                          to_process.delete if to_process
                          GoodData::Process.deploy_from_appstore(process.path, name: process.name, client: to_project.client, project: to_project)
+                       elsif process.component
+                         to_process.delete if to_process
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Calls 'to_process.delete' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/models/project.rb]: Tests 'to_process' at least 4 times (RepeatedConditional)

File: lib/gooddata/models/project.rb [hunk 10]
```diff
@@ -264,6 +264,9 @@ def transfer_processes(from_project, to_project, options = {})
...
+                         GoodData::Process.deploy_component(GoodData::Helpers.symbolize_keys(process.to_hash), project: to_project, client: to_project.client)
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Calls 'to_project.client' 2 times (DuplicateMethodCall)

File: lib/gooddata/models/project.rb [hunk 11]
```diff
@@ -264,6 +264,9 @@ def transfer_processes(from_project, to_project, options = {})
           to_process = if process.path
                          to_process.delete if to_process
                          GoodData::Process.deploy_from_appstore(process.path, name: process.name, client: to_project.client, project: to_project)
+                       elsif process.component
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Refers to 'process' more than self (maybe move it to another class?) (FeatureEnvy)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/89/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/906/ : SUCCESS in 52s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1174/ : SUCCESS in 50s

File: lib/gooddata/models/process.rb [hunk 12]
```diff
@@ -90,16 +90,24 @@ def upload_package(path, files_to_exclude, opts = { :client => GoodData.connecti
       # @option options [String] :name Readable name of the process
       # @option options [String] :process_id ID of a process to be redeployed (do not set if you want to create a new process)
       # @option options [Boolean] :verbose (false) Switch on verbose mode for detailed logging
-      def deploy(path, options = { :client => GoodData.client, :project => GoodData.project })
-        return deploy_brick(path, options) if path.to_s.start_with?(APP_STORE_URL)
-
-        return deploy_from_appstore(path.to_s, options) if (path.to_s =~ %r{\${.*}:(.*)\/(.*):\/}) == 0 # rubocop:disable Style/NumericPredicate
+      def deploy(path, options = { client: GoodData.client, project: GoodData.project })
+        case path
```
Comment (panjan) [inline lib/gooddata/models/process.rb]: We don't need to set `case` if we don't use it in a `when`.

File: lib/gooddata/models/project.rb [hunk 13]
```diff
@@ -404,6 +407,9 @@ def transfer_schedules(from_project, to_project)
               (schedule_spec[:process_id] == local.process_id) && (schedule.name == schedule_spec[:name])
             end
 
+            # processes without schedules
+            # next unless remote_process && process_spec
```
Comment (panjan) [inline lib/gooddata/models/project.rb]: Code commented out.

File: lib/gooddata/models/process.rb [hunk 14]
```diff
@@ -191,26 +181,43 @@ def deploy_from_appstore(path, options = { :client => GoodData.client, :project
...
+        save(data, options)
+      end
+
+      def deploy_component(data, options = { client: GoodData.client, project: GoodData.project })
+        client, project = GoodData.get_client_and_project(options)
+        data = { process: data } unless data[:process]
+        data[:process] = GoodData::Helpers.symbolize_keys(data[:process]).select { |k| [:type, :name, :component].include? k }
+        data[:process][:component] = GoodData::Helpers.symbolize_keys(data[:process][:component]).select { |k| [:name, :version].include? k }
+
+        # TODO: we should not have to do this
```
Comment (panjan) [inline lib/gooddata/models/process.rb]: todo

File: lib/gooddata/models/process.rb [hunk 15]
```diff
@@ -90,16 +90,23 @@ def upload_package(path, files_to_exclude, opts = { :client => GoodData.connecti
...
       # @option options [String] :process_id ID of a process to be redeployed (do not set if you want to create a new process)
       # @option options [Boolean] :verbose (false) Switch on verbose mode for detailed logging
-      def deploy(path, options = { :client => GoodData.client, :project => GoodData.project })
-        return deploy_brick(path, options) if path.to_s.start_with?(APP_STORE_URL)
-
-        return deploy_from_appstore(path.to_s, options) if (path.to_s =~ %r{\${.*}:(.*)\/(.*):\/}) == 0 # rubocop:disable Style/NumericPredicate
+      def deploy(path, options = { client: GoodData.client, project: GoodData.project })
+        if (path.is_a?(Hash) && path[:component])
+          deploy_component path, options
+        elsif path.to_s.start_with?(APP_STORE_URL)
```
Comment (yenkins) [inline lib/gooddata/models/process.rb]: Calls 'path.to_s' 3 times (DuplicateMethodCall)

File: lib/gooddata/models/process.rb [hunk 16]
```diff
@@ -90,16 +90,23 @@ def upload_package(path, files_to_exclude, opts = { :client => GoodData.connecti
       # @option options [String] :name Readable name of the process
       # @option options [String] :process_id ID of a process to be redeployed (do not set if you want to create a new process)
       # @option options [Boolean] :verbose (false) Switch on verbose mode for detailed logging
-      def deploy(path, options = { :client => GoodData.client, :project => GoodData.project })
-        return deploy_brick(path, options) if path.to_s.start_with?(APP_STORE_URL)
-
-        return deploy_from_appstore(path.to_s, options) if (path.to_s =~ %r{\${.*}:(.*)\/(.*):\/}) == 0 # rubocop:disable Style/NumericPredicate
+      def deploy(path, options = { client: GoodData.client, project: GoodData.project })
+        if (path.is_a?(Hash) && path[:component])
```
Comment (yenkins) [inline lib/gooddata/models/process.rb]: Refers to 'path' more than self (maybe move it to another class?) (FeatureEnvy)

Comment (yenkins) [inline lib/gooddata/models/process.rb]: Style/ParenthesesAroundCondition: Don't use parentheses around the condition of an `if`.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/123/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/948/ : FAILURE in 50s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1217/ : SUCCESS in 47s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/124/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/949/ : SUCCESS in 41s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1218/ : SUCCESS in 47s

**Comment by panjan** [comment]
ok to merge

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-merge-compliance-check/50/ : SUCCESS in 14s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/970/ : SUCCESS in 53s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1239/ : SUCCESS in 49s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1394/ : FAILURE in 53m 59s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1159/ : UNSTABLE in 2m 55s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/140/ : SUCCESS in 13s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/982/ : SUCCESS in 52s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1251/ : SUCCESS in 58s

**Comment by panjan** [comment]
Tests run locally because of instability of CI. Merging.
88 examples, 0 failures

**Comment by panjan** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/158/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1415/ : SUCCESS in 54m 11s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1183/ : FAILURE in 35s

**Pull request merged**